### PR TITLE
[SPARK-29012][SQL] Support special timestamp values

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?

Supported special string values for `TIMESTAMP` type. They are simply notational shorthands that will be converted to ordinary timestamp values when read. The following string values are supported:
- `epoch [zoneId]` - `1970-01-01 00:00:00+00 (Unix system time zero)`
- `today [zoneId]` - midnight today.
- `yesterday [zoneId]` -midnight yesterday
- `tomorrow [zoneId]` - midnight tomorrow
- `now` - current query start time.

For example:
```sql
spark-sql> SELECT timestamp 'tomorrow';
2019-09-07 00:00:00
```

### Why are the changes needed?

To maintain feature parity with PostgreSQL, see [8.5.1.4. Special Values](https://www.postgresql.org/docs/12/datatype-datetime.html)

### Does this PR introduce any user-facing change?

Previously, the parser fails on the special values with the error:
```sql
spark-sql> select timestamp 'today';
Error in query: 
Cannot parse the TIMESTAMP value: today(line 1, pos 7)
```
After the changes, the special values are converted to appropriate dates:
```sql
spark-sql> select timestamp 'today';
2019-09-06 00:00:00
```

### How was this patch tested?
- Added tests to `TimestampFormatterSuite` to check parsing special values from regular strings.
- Tests in `DateTimeUtilsSuite` check parsing those values from `UTF8String`
- Uncommented tests in `timestamp.sql`

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/results/pgSQL/timestamp.sql.out
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/results/pgSQL/timestamp.sql.out]: This is the epoch in UTC (`1970-01-01 00:00:00`) displayed in the local time zone.

**Comment by MaxGekk** [comment]
@dongjoon-hyun @cloud-fan @maropu Could you take a look at this when you have time.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: What's different from `convertSpecialDate`? I know the output dataType is different, but the way to handle these special values is different, too?
https://github.com/apache/spark/pull/25708/files#diff-da60f07e1826788aaeb07f295fae4b8aR866
Can we share some code between them?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: How about checking if an input is trimmed by `assert`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala]: nit: remove `{`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: None if a zone id is defined instead of ignoring it?
```
postgres=# select timestamp 'now CET';
ERROR:  invalid input syntax for type timestamp: "now CET"
LINE 1: select timestamp 'now CET';
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: The description for the supported special values in this pr should be;
```
epoch [zoneid] - 1970-01-01 00:00:00+00 (Unix system time zero)
today [zoneid] - midnight today.
yesterday [zoneid] -midnight yesterday
tomorrow [zoneid] - midnight tomorrow
now - current query start time.
```
?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: Can you check illegal cases, e.g., `now CET`

File: sql/core/src/test/resources/sql-tests/inputs/pgSQL/timestamp.sql
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/pgSQL/timestamp.sql]: oh... wrong jira number... that must be SPARK-29012...

File: sql/core/src/test/resources/sql-tests/results/pgSQL/timestamp.sql.out
Comment (maropu) [reply sql/core/src/test/resources/sql-tests/results/pgSQL/timestamp.sql.out]: To print `1970-01-01 00:00:00` here, better to set a config for pgSQL tests in
https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala#L301 ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: None if a zone id is defined instead of ignoring it?
```
postgres=# select timestamp 'now CET';
ERROR:  invalid input syntax for type timestamp: "now CET"
LINE 1: select timestamp 'now CET';
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I have extracted common code there https://github.com/apache/spark/pull/25716/files#diff-da60f07e1826788aaeb07f295fae4b8aR864-R890

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I will add the assert:
```scala
assert(input.trim.length == input.length)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Because I need `String` inside of `extractSpecialValue`, and `UTF8String.fromString` converts `UTF8String` to `String` via `Array[Byte]`. Why should we convert the same string to bytes twice?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Why?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: I have already added the test here https://github.com/apache/spark/pull/25716/files#diff-c5655e947ce2dd3748e4cf95ebc32e8aR580

File: sql/core/src/test/resources/sql-tests/inputs/pgSQL/timestamp.sql
Comment (MaxGekk) [reply sql/core/src/test/resources/sql-tests/inputs/pgSQL/timestamp.sql]: This is because of it was cloned.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I have already fixed that.

File: sql/core/src/test/resources/sql-tests/results/pgSQL/timestamp.sql.out
Comment (MaxGekk) [reply sql/core/src/test/resources/sql-tests/results/pgSQL/timestamp.sql.out]: I can set `UTC` globally but ... if we know the reason of this, should we do that?

Comment (MaxGekk) [reply sql/core/src/test/resources/sql-tests/results/pgSQL/timestamp.sql.out]: And I am afraid it won't be enough. Need to set system time zone as well: https://github.com/apache/spark/blob/7cc0f0e9a795431a79408644b3473df499118d1e/sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala#L552-L553

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I'm not 100% sure about bytecode for this though, no overhead to use `return`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Ur, I see.

File: sql/core/src/test/resources/sql-tests/results/pgSQL/timestamp.sql.out
Comment (maropu) [reply sql/core/src/test/resources/sql-tests/results/pgSQL/timestamp.sql.out]: NVM. I think its ok as it is. But, better to leave some comments about why in `timestamp.sql`.

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @MaxGekk . I support this approach.
One thing I'm concerning is the coordination between this PR and the followings.
- SPARK-28934 Add `spark.sql.compatiblity.mode`
- SPARK-28989 Add `spark.sql.ansi.enabled`
- SPARK-28997 Add `spark.sql.dialect`

Hi, @gatorsmile , @gengliangwang , @cloud-fan . 
- Can we proceed the feature implementation first in the non-critical PR like this?
- Or, do we need to wait and choose one of them to follow?

**Comment by maropu** [comment]
The current code looks ok to me, but this might be blocked by https://github.com/apache/spark/pull/25697 because of the same reason with #25708 ...

**Comment by gengliangwang** [comment]
Well, for this feature it doesn't really conflict with the current Spark behavior. I think we can proceed it.
I will continue the work of #25697 this week.

**Comment by MaxGekk** [comment]
@maropu @dongjoon-hyun Can we continue with the PR or we are waiting for @gengliangwang 's https://github.com/apache/spark/pull/25697 ?

**Comment by MaxGekk** [comment]
@dongjoon-hyun @maropu Can you merge this PR? Checking the flag could be added in #25697 itself or in a separate follow-up PR after the merge of #25697, I do believe.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
@MaxGekk so do you plan to hide this change behind the configuration `spark.sql.dialect` later?

**Comment by MaxGekk** [comment]
> so do you plan to hide this change behind the configuration spark.sql.dialect later?

@HyukjinKwon Yes, I do. I would do that in a separate PR as soon as the flag is available in the master branch.

**Review by HyukjinKwon — APPROVED** [review]
LGTM but please @MaxGekk make sure adding a configuration later. Otherwise we might have to revert before 3.0

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by MaxGekk** [comment]
This PR https://github.com/apache/spark/pull/25834 hide the feature under the SQL config `spark.sql.dialect = "PostgreSQL"`

**Pull request closed without merging**
